doc/ip/next_hop_group_hld.md

```diff
@@ -0,0 +1,148 @@
+
+A more efficient system would involve managing the next hop groups in use by the route table separately, and simply have the route table specify a reference to which next hop group to use.  Since at scale many routes will use the same next hop groups, this requires much smaller occupancy per route, and so more efficient building, transmission and parsing of per-route information.
+
+# 2 Requirement Overview
+## 2.1 Functional Requirements
+
+- Improve efficiency of storage of routing information in APP_DB.
+- Reduce time to reprogram routes when next hop information changes.
+
+## 2.2 Configuration and Management Requirements
+- The existing 'show ip route' and 'show ipv6 route' commands must be supported when using the new next hop group table.
+
+## 2.3 Scalability Requirements
+- Unchanged in requirements, but will improve scalability in terms of storage and programming time.
+
+## 2.4 Warm Boot Requirements
+- Unchanged - the new next hop group table must be compatible with existing warm boot requirements.
+
+# 3 Design
+## 3.1 Overview
+This design directly changes APP_DB and orchagent, along with minor changes to support the existing CLI when using the new APP_DB table.
+
+## 3.2 DB Changes
+### 3.2.1 APP DB
+Currently the ROUTE_TABLE in APP_DB includes all next hop information for the route, as follows:
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+ ```
+
+This design adds a new NEXT_HOP_GROUP_TABLE, to store next hop group information to be used by one or more routes.
+```
+### NEXT_HOP_GROUP_TABLE
+    ;Stores a list of groups of one or more next hops
+    ;Status: Mandatory
+    key           = NEXT_HOP_GROUP_TABLE:string ; arbitrary identifier for the next hop group
+    nexthop       = *prefix,           ; IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+```
+The ROUTE_TABLE is then extended to allow a reference to the next hop group to be specified, instead of the current nexthop and intf fields.
+```
+### ROUTE_TABLE
+    ;Stores a list of routes
+    ;Status: Mandatory
+    key           = ROUTE_TABLE:prefix
+    nexthop       = *prefix, ;IP addresses separated “,” (empty indicates no gateway)
+    ifname        = *PORT_TABLE.key,   ; zero or more separated by “,” (zero indicates no interface)
+    blackhole     = BIT ; Set to 1 if this route is a blackhole (or null0)
+    nexthop_group = NEXT_HOP_GROUP_TABLE:key ; index within the NEXT_HOP_GROUP_TABLE, optionally used instead of nexthop and intf fields
+ ```
+
+## 3.3 Switch State Service Design
+### 3.3.1 Orchestration Agent
```

Comment:
Can you please add code flow for the nexthop group orch? Say if nexthop group has only one nexthop, does it translate to SAI ECMP group with one member or is it single nexthop..

Reply:
Done